File: src-java/src/main/java/QuickSort.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * ClassName: QuickSort
 * Package: PACKAGE_NAME
 * Description:
 *
 * @Author AnXin
 * @Create 2026/5/27 22:18
 * @Version 1.0
 */
public class QuickSort {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int n = scan.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = scan.nextInt();
        }
        quickSort(arr, 0, arr.length - 1);
        System.out.println(Arrays.toString(arr));
    }

    static void quickSort(int[] q, int l, int r) {
        if (l >= r) {
            return;
        }

        int i = l - 1;
        int j = r + 1;
        int x = q[l + r >> 1];
        while (i < j) {
            do {
                i++;
            } while (q[i] < x);
            do {
                j--;
            } while (q[j] > x);
            if (i < j) {
                int temp = q[i];
                q[i] = q[j];
                q[j] = temp;
            }
        }
        quickSort(q, l, j);
        quickSort(q, j + 1, r);
    }
}

File: src-python/quick_sort.py
def quick_sort(q: list[int], l: int, r: int) -> None:
    if l >= r:
        return

    i = l - 1
    j = r + 1
    x = q[l + r >> 1]
    while i < j:
        i = i + 1
        while q[i] < x:
            i = i + 1
        j = j - 1
        while q[j] > x:
            j = j - 1
        if i < j:
            q[i], q[j] = q[j], q[i]
    quick_sort(q, l, j)
    quick_sort(q, j + 1, r)

if __name__ == '__main__':
    n = int(input())
    arr = list(map(int, input().split()))
    quick_sort(arr, 0, len(arr) - 1)
    print(arr)